### PR TITLE
FMA | Javascript Client | Library Structure

### PR DESCRIPTION
**Pull request by openrijal**
- directory structure and base files for javascript client library
- directory structure for example javascript project folder

### CODE REVIEW

File: examples/client_examples/javascript_client/README.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+1. Install requirements  
+*** Note if `npm setup` ran from top level of repo please skip step 1***
+```console
+npm install 
+```
+
+2. EXAMPLE
+```console
+bash run_example.sh 3 abc123 average_layers http://host.docker.internal:8000 http://0.0.0.0:8080
```
Comment (JGSweets) [inline examples/client_examples/javascript_client/README.md:13]: Let's update `abc123` to be `<user_id>`

File: clients/javascript_client/fma_connect/clients.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+import { DefaultSettings } from './settings';
+
+const defaultSettings = new DefaultSettings()
```
Comment (ksneab7) [inline clients/javascript_client/fma_connect/clients.ts:3]: Ensure that the files end with proper spacing

**Comment by JGSweets** [comment]
@ksneab7 @taylorfturner bump for review

File: clients/javascript_client/fma_connect/settings.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+export class DefaultSettings {
+    url: string;
+    constructor(){
+        this.url = "http://127.0.0.1:8000/";
+    }
+}
```
Comment (taylorfturner) [inline clients/javascript_client/fma_connect/settings.ts:6]: new line EOF

File: clients/javascript_client/jest.config.js [hunk 4]
```diff
@@ -0,0 +1,5 @@
+/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
+module.exports = {
+    preset: 'ts-jest',
+    testEnvironment: 'node',
+  };
```
Comment (taylorfturner) [inline clients/javascript_client/jest.config.js:5]: new line EOF

File: clients/javascript_client/package.json [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    "node-fetch": "^2.6.1",
+    "argparse": "^2.0.1",
+    "ts-node": "^10.8.1"
+  },
+  "scripts": {
+    "compile": "rm -rfv dist/lib && tsc && tsc --build tsconfig.es5.json",
+    "test": "jest --coverage",
+    "lint": "eslint fma_connect --ext .ts"
+  }
+}
```
Comment (taylorfturner) [inline clients/javascript_client/package.json:28]: new line EOF

File: clients/javascript_client/tsconfig.json [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+    "compilerOptions": {
+        "module": "ES6",
+        "target": "ES5",
+        "lib": ["ES2020", "DOM"],
+        "declaration": true,
+        "outDir": "./dist/lib/es6",
+        "moduleResolution": "node"
+    },
+    "include": ["fma_connect/**/*"]
+}
```
Comment (taylorfturner) [inline clients/javascript_client/tsconfig.json:11]: new line EOF

File: examples/client_examples/javascript_client/package.json [hunk 7]
```diff
@@ -0,0 +1,9 @@
+{
+  "dependencies": {
+    "@tensorflow/tfjs": "^3.18.0",
+    "http-server": "^14.1.1",
+    "ts-node": "^10.8.1",
+    "argparse":"^3.9.0"
+  }
+}
+
```
Comment (taylorfturner) [inline examples/client_examples/javascript_client/package.json:9]: same

File: examples/client_examples/javascript_client/utils.ts [hunk 8]
```diff
@@ -0,0 +1,3 @@
+export const pluck = (elements, field) => {
+    return elements.map((element) => element[field])
+}
```
Comment (taylorfturner) [inline examples/client_examples/javascript_client/utils.ts:3]: same

File: clients/javascript_client/README.md [hunk 9]
```diff
@@ -1 +1,15 @@
 # Javascript Client for FMA service
+An api client wrapper for connecting to the federated-learning-service.
+
+## Setup
+Please run the ```setup``` at the top level of the directory
```
Comment (taylorfturner) [inline clients/javascript_client/README.md:5]:
```suggestion
Please run the the following command at the top level of the repo
```
currently which directory is ambiguous...  is it the `clients/` or `javascript_client` or entire repo?

File: clients/javascript_client/README.md [hunk 10]
```diff
@@ -1 +1,15 @@
...
+```npm setup```
+
+### Example Setup
+[README.md](examples/browser_client/README.md)
+
+## Testing
+### npm test
+To run testing of the API interface functions run  
+```npm test```
```
Comment (taylorfturner) [inline clients/javascript_client/README.md:14]: any particular location this needs to be run?

File: examples/client_examples/javascript_client/README.md [hunk 11]
```diff
@@ -0,0 +1,19 @@
+# In-Browser Training Example
+
+Quickstart --
```
Comment (taylorfturner) [inline examples/client_examples/javascript_client/README.md:3]:
```suggestion
Quickstart
```

Comment (taylorfturner) [reply examples/client_examples/javascript_client/README.md:3]: not sure the `--` is needed

File: examples/client_examples/javascript_client/README.md [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+
+1. Install requirements  
+*** Note if `npm setup` ran from top level of repo please skip step 1***
+```console
+npm install 
+```
+
+2. EXAMPLE
```
Comment (taylorfturner) [inline examples/client_examples/javascript_client/README.md:11]:
```suggestion
2. Example Run
```

File: examples/client_examples/javascript_client/README.md [hunk 13]
```diff
@@ -0,0 +1,19 @@
+# In-Browser Training Example
+
+Quickstart --
+
+1. Install requirements
```
Comment (taylorfturner) [inline examples/client_examples/javascript_client/README.md:5]:
```suggestion
1. Install  
```

File: examples/client_examples/javascript_client/README.md [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+```console
+npm install 
+```
+
+2. EXAMPLE
+```console
+bash run_example.sh 3 <user_id> average_layers http://host.docker.internal:8000 http://0.0.0.0:8080
+```
+
+3. Run Client.ts
```
Comment (taylorfturner) [inline examples/client_examples/javascript_client/README.md:16]:
```suggestion
3. Run Client (`client.ts`)
```

**Review by taylorfturner — CHANGES_REQUESTED** [review]
Looking good -- some new line at EOF and verbiage. Otherwise, looking good

File: clients/javascript_client/jest.config.js [hunk 15]
```diff
@@ -0,0 +1,6 @@
...
+
```
Comment (taylorfturner) [inline clients/javascript_client/jest.config.js:6]: you may have some unneeded white space here

File: examples/client_examples/javascript_client/utils.ts (hunk 8 above)
Comment (taylorfturner) [reply examples/client_examples/javascript_client/utils.ts:3]: new line EOF

File: clients/javascript_client/README.md [hunk 16]
```diff
@@ -1 +1,12 @@
 # Javascript Client for FMA service
+An api client wrapper for connecting to the federated-learning-service.
+
+## Setup
+Please run the `npm setup` at the top level of the directory
+
+### Example Setup
+[README.md](examples/browser_client/README.md)
```
Comment (taylorfturner) [inline clients/javascript_client/README.md:8]: this appears to be a non-existent path...

**Review by taylorfturner — CHANGES_REQUESTED** [review]

File: clients/javascript_client/README.md [hunk 17]
```diff
@@ -1 +1,12 @@
 # Javascript Client for FMA service
+An api client wrapper for connecting to the federated-learning-service.
+
+## Setup
+Please run `npm setup` at `clients/javascript_client` directory
```
Comment (taylorfturner) [inline clients/javascript_client/README.md:5]:
```suggestion
Run `npm setup` in `clients/javascript_client` directory
```

File: clients/javascript_client/README.md [hunk 18]
```diff
@@ -1 +1,12 @@
...
+
+### Example Setup
+[README.md](examples/client_examples/javascript_client/README.md)
```
Comment (taylorfturner) [inline clients/javascript_client/README.md:8]: maybe render the other README here... so the user doesn't have to open another window

**Review by taylorfturner — CHANGES_REQUESTED** [review]

File: clients/javascript_client/README.md (hunk 18 above)
Comment (taylorfturner) [reply clients/javascript_client/README.md:8]: ah I see

File: examples/client_examples/javascript_client/README.md [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+
+Quickstart
+
+1. Install  
+*** Note if `npm setup` ran from top level of repo please skip step 1***
+```console
+npm install 
+```
+
+2. Example Run
```
Comment (taylorfturner) [inline examples/client_examples/javascript_client/README.md:11]:
```suggestion
2. Run Example
```
re-word to match the `bash` script that is actually run

**Review by taylorfturner — CHANGES_REQUESTED** [review]

**Review by taylorfturner — APPROVED** [review]
LGTM

**Comment by taylorfturner** [comment]
bump @tyfarnan @micdavis

**Pull request merged**